### PR TITLE
Fix: HttpMethodMatcherPolicy.ApplyAsync ArgumentNullException

### LINKED ISSUE #39398: HttpMethodMatcherPolicy.ApplyAsync ArgumentNullException
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

TL;DR: https://github.com/dotnet/aspnetcore/blob/077d0883e943bebbe8151ead202d4c18cc3bee6b/src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs#L165 can throw ArgumentNullException (note the `!`) - the OrderBy should probably be skipped altogether if `methods` is null.

While tweaking my integration test infra to allow running the tests in parallel, I ran into a weird issue arising every now and then when running tests in a particular order. Not going into too much detail, but the issue reproduces if I run a specific set of tests in parallel. The issue can, weirdly enough, be circumvented by making a seemingly useless API call via the test server before running the actual test.

The stack trace:
```
System.ArgumentNullException
Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.OrderedEnumerable`2..ctor(IEnumerable`1 source, Func`2 keySelector, IComparer`1 comparer, Boolean descending, OrderedEnumerable`1 parent)
   at System.Linq.Enumerable.OrderBy[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IComparer`1 comparer)
   at Microsoft.AspNetCore.Routing.Matching.HttpMethodMatcherPolicy.ApplyAsync(HttpContext httpContext, CandidateSet candidates)
   at Microsoft.AspNetCore.Routing.Matching.DfaMatcher.SelectEndpointWithPoliciesAsync(HttpContext httpContext, IEndpointSelectorPolicy[] policies, CandidateSet candidateSet)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatch|8_1(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task matchTask)
   at Microsoft.AspNetCore.TestHost.HttpContextBuilder.<>c__DisplayClass23_0.<<SendAsync>g__RunRequestAsync|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.TestHost.ClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Testing.Handlers.CookieContainerHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
```

### Expected Behavior

The line of code in question should not expect `methods` to be not null

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version

6.0.100

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by z77ma**
# Fix of a possible ArgumentNullException
Applied an approach used in the method 'GetEdges' with inlined collection initialization instead of deferred approach.

## Description
The 'ArgumentNullException' is thrown by execution of the 'OrderBy' Linq method over uninitialized collection (local variable 'methods'). This situation is possibile under following conditions:

* all candidates passed to the 'ApplyAsync' method are invalid (line 123)
* **and** endpoints of the candidates **either** don't contain metadata **or** 'HttpMethod' collection within them is empty (line 113)

As a solution I suggest to initialize the 'methods' collection in place of defining it.

Alternatives might be:
* checking the 'methods' collection against null before executing the 'OrderBy' extension method
* refusing from ordering the colection

Unit test to cover the above case was added.

Fixes #39398

### CODE REVIEW

**Comment by z77ma** [comment]
@pranavkm, saw your comment in another PR proposal. Would you consider this change?

File: src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs [hunk 1]
```diff
@@ -143,7 +143,6 @@ public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
                 var candidateMethod = metadata.HttpMethods[j];
                 if (!HttpMethods.Equals(httpMethod, candidateMethod))
                 {
-                    methods = methods ?? new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                     methods.Add(candidateMethod);
```
Comment (pranavkm) [inline src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:146]:
```suggestion
                    methods = methods ?? new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                    methods.Add(candidateMethod);
```

File: src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs [hunk 2]
```diff
@@ -162,7 +161,7 @@ public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
         if (needs405Endpoint == true)
         {
             // We saw some endpoints coming in, and we eliminated them all.
-            httpContext.SetEndpoint(CreateRejectionEndpoint(methods!.OrderBy(m => m, StringComparer.OrdinalIgnoreCase)));
+            httpContext.SetEndpoint(CreateRejectionEndpoint(methods.OrderBy(m => m, StringComparer.OrdinalIgnoreCase)));
```
Comment (pranavkm) [inline src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs:164]:
```suggestion
            httpContext.SetEndpoint(CreateRejectionEndpoint(methods?.OrderBy(m => m, StringComparer.OrdinalIgnoreCase)));
```

**Comment by pranavkm** [comment]
@z77ma an alternative would be to update `CreateRejectionEndpoint` to allow a null `methods` argument. I updated your PR to include a couple of suggestions.

File: src/Http/Routing/test/UnitTests/Matching/HttpMethodMatcherPolicyTest.cs [hunk 3]
```diff
@@ -151,6 +151,28 @@ public void IEndpointSelectorPolicy_AppliesToNode_EndpointIsNotDynamic_ReturnsFa
...
     }
 
+    [Theory]
+    [InlineData(1)]
+    [InlineData(2)]
+    public async Task IEndpointSelectorPolicy_ApplyAsync_ProcessesInvalidCandidate(int candidateNum)
+    {
+        var policy = (IEndpointSelectorPolicy)CreatePolicy();
+
+        RouteEndpoint[] endpoints = new RouteEndpoint[candidateNum];
```
Comment (pranavkm) [inline src/Http/Routing/test/UnitTests/Matching/HttpMethodMatcherPolicyTest.cs:161]:
```suggestion
        var endpoints = new RouteEndpoint[candidateNum];
```

**Comment by z77ma** [comment]
@pranavkm, thanks, this looks even better.
Out of curiosity, what would you see as an advantage in deferring the collection initialization? Is a chance, that it won't be even needed, that high?

**Comment by pranavkm** [comment]
The routing code is considered perf sensitive, so removing any extraneous allocations in the default code path where everything succeeds is preferred.

**Review by pranavkm — APPROVED** [review]

**Pull request merged**